src/HonzaBotner.Discord.Services/Reactions/ZiraHandler.cs

```diff
@@ -0,0 +1,71 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using DSharpPlus.Entities;
+using DSharpPlus.EventArgs;
+using HonzaBotner.Services.Contract;
+
+namespace HonzaBotner.Discord.Services.Reactions
+{
+    public class ZiraHandler : IReactionHandler
+    {
+        private readonly IZiraService _ziraService;
+
+        public ZiraHandler(IZiraService ziraService)
+        {
+            _ziraService = ziraService;
+        }
+
+        public async Task<IReactionHandler.Result> HandleAddAsync(MessageReactionAddEventArgs eventArgs)
+        {
+            ICollection<ulong> mappings =
+                await _ziraService.FindMappingAsync(eventArgs.Channel.Id, eventArgs.Message.Id, eventArgs.Emoji.Name);
+            if (!mappings.Any())
+                return IReactionHandler.Result.Continue;
+
+            foreach (ulong roleId in mappings)
+            {
+                DiscordRole? role = GetRole(eventArgs.Guild, roleId);
+                if(role == null)
+                    continue;
+
+                DiscordMember member = await GetMemberAsync(eventArgs.Guild, eventArgs.User.Id);
+
+                await member.GrantRoleAsync(role, "Zira");
+            }
+
+            return IReactionHandler.Result.Stop;
+        }
+
+        public async Task<IReactionHandler.Result> HandleRemoveAsync(MessageReactionRemoveEventArgs eventArgs)
+        {
+            ICollection<ulong> mappings =
+                await _ziraService.FindMappingAsync(eventArgs.Channel.Id, eventArgs.Message.Id, eventArgs.Emoji.Name);
+            if (!mappings.Any())
+                return IReactionHandler.Result.Continue;
+
+            foreach (ulong roleId in mappings)
+            {
+                DiscordRole? role = GetRole(eventArgs.Guild, roleId);
+                if(role == null)
+                    continue;
+
+                DiscordMember member = await GetMemberAsync(eventArgs.Guild, eventArgs.User.Id);
+
+                await member.RevokeRoleAsync(role, "Zira");
```

Comment:
Might throw an exception.

Reply:
Line 49: Ok
Line 53: Not sure
Line 55: Really?

Reply:
![image](https://user-images.githubusercontent.com/5287596/108560423-fb65a600-72fc-11eb-8569-1d896e920055.png)

Reply:
If it throws exception we will handle it in DiscordBot